README.md
# Cascaded Sobol' Sampling

Code of the paper: [Cascaded Sobol' Sampling](https://projet.liris.cnrs.fr/cascaded/paper.pdf), Loïs Paulin, David Coeurjolly, Jean-Claude Iehl, Nicolas Bonneel, Alexander Keller, Victor Ostromoukhov, ACM Transactions on Graphics (Proceedings of SIGGRAPH Asia), 40(6), pp. 274:1–274:13, December 2021.

![](https://projet.liris.cnrs.fr/cascaded/teaser.png)


Building
========

To build project:

```
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
make
```

Running
=======

Generate a point set of 16 points in 3 dimension in file `owenCascadedSobol_3D_16pts.dat` using `cascaded_sobol_init_tab.dat` as sobol generator matrices and with owen scrambling

    ./cascadedSobol -d 3 -n 16 -i ../data/cascaded_sobol_init_tab.dat -o owenCascadedSobol_3D_16pts.dat -p

License
=======

```
 MIT License
 
 Copyright (c) 2021 CNRS
 
 Loïs Paulin, David Coeurjolly, Jean-Claude Iehl, Nicolas Bonneel, Alexander Keller, Victor Ostromoukhov
 "Cascaded Sobol' Sampling", ACM Transactions on Graphics (Proceedings of SIGGRAPH Asia), 40(6), pp. 274:1–274:13, December 2021
 December 2021
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 ``` 


Changelog.md
* 2023-12: fix duplicate symbol error and emscripten build error (@wkjarosz, Wojciech Jarosz, [#1](https://github.com/loispaulin/CascadedSobol/pull/1))
* 2023-06: Initial commits with the source code of the siggraph paper (David Coeurjolly)


Samplers/OwenScrambling.h
/**
 MIT License
 
 Copyright (c) 2021 CNRS
 
 Loïs Paulin, David Coeurjolly, Jean-Claude Iehl, Nicolas Bonneel, Alexander Keller, Victor Ostromoukhov
 "Cascaded Sobol' Sampling", ACM Transactions on Graphics (Proceedings of SIGGRAPH Asia), 40(6), pp. 274:1–274:13, December 2021
 December 2021
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
#ifndef SOBOL_OWENSCRAMBLING_H
#define SOBOL_OWENSCRAMBLING_H

#include <bitset>
#include <iostream>
#include "Random.h"

//seed has to stay the same for a given dimention
uint32_t OwenScrambling(uint32_t sobolPoint, uint32_t seed, const uint32_t owen_tree_depth) {

    RNG scramble_rng;

    constexpr uint32_t bits= 8*sizeof(uint32_t);
    constexpr uint32_t shift= 8*sizeof(uint32_t) -1;

    // flip root, node_index == 0, implicit rng.index(0)
    scramble_rng.seed(seed);
    uint32_t flip= scramble_rng.sample_range(2) << shift;
    uint32_t code= sobolPoint ^ flip;        // flip MSB

    for(uint32_t idigit= 1; idigit < owen_tree_depth; idigit++)
    {
        uint32_t level_base= (1u << idigit) - 1;
        uint32_t level_offset = sobolPoint >> (bits - idigit);	// level > 0 == 2^d nodes
        uint32_t node_index = level_base + level_offset;

        uint32_t node_seed= seed ^ node_index;
        scramble_rng.seed(node_seed);

        uint32_t flip= scramble_rng.sample_range(2) << (shift - idigit);
        code= code ^ flip;
    }

    return code;
}


#endif //SOBOL_OWENSCRAMBLING_H


Samplers/SobolGenerator1D.h
/**
 MIT License
 
 Copyright (c) 2021 CNRS
 
 Loïs Paulin, David Coeurjolly, Jean-Claude Iehl, Nicolas Bonneel, Alexander Keller, Victor Ostromoukhov
 "Cascaded Sobol' Sampling", ACM Transactions on Graphics (Proceedings of SIGGRAPH Asia), 40(6), pp. 274:1–274:13, December 2021
 December 2021
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
#ifndef MYSOBOL_SOBOLGENERATOR1D_H
#define MYSOBOL_SOBOLGENERATOR1D_H

#include <vector>
#include <array>
#include <cmath>
#include <iostream>
#include <sstream>
#include <bitset>
#include <climits>
#include <limits>
#include <sstream>
#include <fstream>

inline uint32_t ReverseBits(uint32_t n) {
    n = (n << 16) | (n >> 16);
    n = ((n & 0x00ff00ff) << 8) | ((n & 0xff00ff00) >> 8);
    n = ((n & 0x0f0f0f0f) << 4) | ((n & 0xf0f0f0f0) >> 4);
    n = ((n & 0x33333333) << 2) | ((n & 0xcccccccc) >> 2);
    n = ((n & 0x55555555) << 1) | ((n & 0xaaaaaaaa) >> 1);
    return n;
}

class SobolGenerator1D{

public:
    //dimension of Sobol polynomial
    uint32_t s;
    uint32_t d;

    //binary Sobol polynomial
    uint32_t a;

    //Sobol matrice
    std::array<uint32_t, CHAR_BIT*sizeof(uint32_t)> m;

    inline SobolGenerator1D(){}

    inline SobolGenerator1D(uint32_t _d, uint32_t _s, uint32_t _a, const std::vector<uint32_t>& _m){
        init1D(_d, _s, _a, _m);
    }

    //Generate matrix from sobol polynomial rule
    inline void generateMatrix() {
        //fill each matrix rank
    	if (s == 0) {	// van der Corput
    		for (uint32_t k = s; k < m.size(); ++k)
    			m[k] = 1;
    	} else {
            for (uint32_t k = s; k < m.size(); ++k) {
                m[k] = 0;
                for (int i = 1; i < s; ++i) {
                    // `akj` stores aj_k, note that the polynomial rep is reversed
                    // `pw2` stores 2^k
                    const uint32_t ai = (a >> (s - 1 - i)) & 1;
                    const uint32_t pw2 = (1 << i);
                    m[k] ^= ai * pw2 * m[k - i];
                }
                m[k] ^= (1 << s) * m[k - s];
                m[k] ^= m[k - s];
            }
    	}
    }

    inline void init1D(uint32_t _d, uint32_t _s, uint32_t _a, const std::vector<uint32_t>& _m) {
        d = _d;
        s = _s;
        a = _a;
        for (int i = 0; i < _m.size(); ++i) {
            m[i] = _m[i];
        }
        generateMatrix();
    }

    inline uint32_t getSobolInt(uint32_t n) const{
        uint32_t res = 0;
        for (int i = 0; i < CHAR_BIT * sizeof(uint32_t); ++i) {
            res <<= 1;
            res ^= (n & 1) * m[i];
            n /= 2;
        }
        return res;
    }

    inline double getSobolDouble(uint32_t n) const{

        uint32_t res = getSobolInt(n);
        return double(res) / std::numeric_limits<uint32_t>::max();
    }

    inline friend std::ostream& operator<<(std::ostream& out, const SobolGenerator1D& sobol){

        out << sobol.d << "\t" << sobol.s << "\t" << sobol.a  << "\t";
        for (uint32_t i = 0; i < sobol.s; ++i) {
            out << sobol.m[i] << " ";
        }

        return out;
    }

};


inline bool readOneInitVector(std::istream& in, uint32_t& d, uint32_t& s, uint32_t& a, std::vector<uint32_t>& m){

    std::string line;
    m.clear();
    //Read a line and skip if it's a header
    bool test;
    do {
        test = bool(getline(in, line));
    } while (test && line[0] == 'd');

    if(test){
        std::istringstream lineStream(line);

        lineStream >> d  >> s >> a;

        uint32_t mi;
        while(lineStream >> mi){
            m.push_back(mi);
        }
        return true;
    } else {
        return false;
    }
}

inline void load_init_table(std::istream& in,
                            std::vector<uint32_t>& d,
                            std::vector<uint32_t>& s,
                            std::vector<uint32_t>& a,
                            std::vector<std::vector<uint32_t>>& m,
                            uint32_t nbMax = -1,
                            uint32_t offset = 0){
    uint32_t ai;
    uint32_t di;
    uint32_t si;
    std::vector<uint32_t> mi;
    std::vector<uint32_t> mi_vdc = {1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1, 1,1};
    while( offset-- && readOneInitVector(in, di, si, ai, mi) );
    int i = 0;
    // first, push a pseudo-entry for dim 0 == van der corput
    d.push_back(0);
    s.push_back(0);
    a.push_back(0);
    m.push_back(std::move(mi_vdc));
    // now, read enties from the input file
    while( i++ != nbMax && readOneInitVector(in, di, si, ai, mi) ){
        d.push_back(di);
        s.push_back(si);
        a.push_back(ai);
        if(si == 0)
        	m.push_back(std::move(mi_vdc));
        else
        	m.push_back(std::move(mi));
   }
}

inline void init_sobols(std::vector<SobolGenerator1D> &sobols,
                        const std::vector<uint32_t>& d,
                        const std::vector<uint32_t>& s,
                        const std::vector<uint32_t>& a,
                        const std::vector<std::vector<uint32_t>>& m
    ) {
    sobols.resize(m.size());
    for(int i = 0; i < m.size(); ++i) {
        sobols[i].init1D(d[i], s[i], a[i], m[i]);
    }
}

#endif //MYSOBOL_SOBOLGENERATOR1D_H


cascadedSobol.cpp
/**
 MIT License
 
 Copyright (c) 2021 CNRS
 
 Loïs Paulin, David Coeurjolly, Jean-Claude Iehl, Nicolas Bonneel, Alexander Keller, Victor Ostromoukhov
 "Cascaded Sobol' Sampling", ACM Transactions on Graphics (Proceedings of SIGGRAPH Asia), 40(6), pp. 274:1–274:13, December 2021
 December 2021
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
#include <iostream>
#include <string>
#include <fstream>
#include <random>

#include "CLI11.hpp"

#include "Samplers/SobolGenerator1D.h"
#include "Samplers/OwenScrambling.h"


using namespace std;

template< typename T >
inline void getCascadedSobol( T *values,
                const std::vector<SobolGenerator1D >& sobols,
                const uint32_t *seeds,
                const int nDims,
                const uint32_t n,
                const uint32_t nbits,
                const bool owen_permut_flag = true,
                const uint32_t owen_tree_depth = 32 )
 {
    uint32_t index = n;		// dim 0: take n-th point as index -> into 32-bit integer
    for(int idim = 0; idim < nDims; idim++)
    {
        index = sobols[idim].getSobolInt(index);	// radix-inversion + sobol
        uint32_t result = index;
        if(owen_permut_flag)						// we need to apply OwenScrambling only when this flag is set
            result = OwenScrambling(result, seeds[idim], owen_tree_depth);
        values[idim] = T(result) / T(std::numeric_limits<uint32_t>::max());	// final value (double) for this dimension
        index = index >> (32-nbits);				// this will be used as new index for the next dimension
    }
}   // getCascadedSobol

int main(int argc, char **argv) {

    //Call parameters handling
    CLI::App app { "cascadedSobol" };
    int nDims = 4;
    app.add_option("-d,--nDims", nDims, "number of dimensions to generate, default: " + std::to_string(nDims)  );
    int npts = 1024;
    app.add_option("-n,--npts", npts, "number of points to generate, default: " + std::to_string(npts) );
    int seed = 133742;
    app.add_option("-s,--seed", seed, "Random number generator seed. default: " + std::to_string(seed)  );
    string input_dir_vectors = "../data/cascaded_sobol_init_tab.dat";
    app.add_option("-i,--idv", input_dir_vectors, "input sobol initialisation (ascii file), default: " + input_dir_vectors );
    bool owen_permut_flag = false;
    app.add_flag("-p", owen_permut_flag, "Apply Owen permutation on output points, default: " + std::to_string(owen_permut_flag) );
    int nbReal = 1;
    app.add_option("-m", nbReal, "number of realizations of the sampler, default: " + std::to_string(nbReal));
    std::string output_fname = "out.dat";
    app.add_option("-o,--output", output_fname, "output fname, default: " + output_fname );
    CLI11_PARSE(app, argc, argv)

    //Create sobol matrices data
    std::vector<SobolGenerator1D > sobols;	// array of sobol data per dim
    std::vector<uint32_t> d;
    std::vector<uint32_t> s;
    std::vector<uint32_t> a;
    std::vector<std::vector<uint32_t>> m;

    //Read sobol matrices value from file
    std::ifstream tableFile(input_dir_vectors);
    if (!tableFile.is_open()) {
        std::cerr << input_dir_vectors << " cannot be read." << std::endl;
        exit(1);
    };
    load_init_table(tableFile, d, s, a, m, nDims);
    init_sobols(sobols, d, s, a, m);

    //Open output file
    std::ofstream ofs(output_fname , std::ios_base::out);

    //init random generator for owen scrambling
    std::uniform_int_distribution<uint32_t> unif32bits(0);  // uniform distribution of uint32_ts between 0 and 2^32-1
    mt19937_64 gen(seed);

    auto nbits = (uint32_t) ((double) log((double) npts) / (double) log((double) 2));
    vector<double> points(npts * nDims);
    //For each realizations
    for(int realization=0; realization < nbReal; ++realization )
    {
        //init owen scrambling seeds
        vector<uint32_t> realSeeds(nDims);
        for (int iDim = 0; iDim < nDims; ++iDim) {
            realSeeds[iDim] = unif32bits(gen);
        }

        //get each point
        for (int ipt = 0; ipt < npts; ipt++) {
            getCascadedSobol(points.data() + ipt * nDims, sobols, realSeeds.data(), nDims, ipt, nbits, owen_permut_flag);
        }

        //export points in ASCII
        for (int ipt = 0; ipt < npts; ipt++) {
            for (unsigned int idim = 0; idim < nDims; idim++) {
                ofs << std::setprecision(20) << std::fixed << points[ipt * nDims + idim] << " ";  // dims {0, ..., nDims-1}
            }
            ofs << endl;
        }
        //Diffent realizations are separated by #
        if (realization+1 != nbReal){
            ofs<<"#"<<std::endl;
        }
    }
    ofs.close();
}

